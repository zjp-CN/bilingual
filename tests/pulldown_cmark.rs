use insta::assert_yaml_snapshot;
use pulldown_cmark::{CowStr, Event, Parser, Tag};

pub fn filter(event: &Event, include: &mut bool) -> bool {
    match event {
        Event::Start(Tag::CodeBlock(_)) => {
            *include = false;
            false
        }
        Event::End(Tag::CodeBlock(_)) => {
            *include = true;
            false
        }
        Event::Start(Tag::Heading(_)) | Event::Start(Tag::Paragraph) => false,
        _ => *include,
    }
}

pub fn extract<'a>(event: Event<'a>) -> CowStr<'a> {
    match event {
        Event::End(Tag::Heading(_)) | Event::End(Tag::Paragraph) => "\n".into(),
        // Event::Text(x) | Event::Code(x) => x,
        Event::Text(x) => x,
        Event::Code(x) => format!("\n{}\n", x).into(),
        _ => " ".into(),
    }
}

#[test]
fn default_config() -> std::io::Result<()> {
    let md = std::fs::read_to_string("assets/8_6_io_eventqueue.md")?;
    assert_yaml_snapshot!("8_6_io_eventqueue", Parser::new(&md).collect::<Vec<_>>());
    assert_yaml_snapshot!("8_6_io_eventqueue_offset",
                          Parser::new(&md).into_offset_iter().collect::<Vec<_>>());

    let parsed = {
        let mut include = true;
        Parser::new(&md).filter(|event| filter(event, &mut include))
                        .map(extract)
                        .collect::<Vec<_>>()
    };
    // assert_yaml_snapshot!("8_6_io_eventqueue_modified", Parser::new(&md).collect::<Vec<_>>());
    assert_yaml_snapshot!("8_6_io_eventqueue_modified", parsed.join(""));

    assert_yaml_snapshot!("8_6_io_eventqueue_translated",
        Parser::new(
            &std::fs::read_to_string("assets/8_6_io_eventqueue_translated.md")?
        ).collect::<Vec<_>>()
    );

    Ok(())
}
