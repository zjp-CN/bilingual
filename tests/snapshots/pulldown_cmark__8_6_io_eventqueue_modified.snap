---
source: tests/pulldown_cmark.rs
expression: "parsed.join(\"\")"

---
"I/O event queue\nThe I/O event queue is what handles most of our I/O tasks. Now we'll go through how we register events to that queue later on, but once an event is ready we it sends the event_id through our channel.\nAs we'll see later on, the way we designed this, we actually made our event_id and callback_id the same value since both represents an unique value for this event. It simplifies things slightly for us.\nWe add the callback_id to the collection of callbacks to run. We pass in Js::Undefined since we'll not actually pass any data along here. You'll see why when we reach the [Http module](./8_3_http_module.md) chapter, but the main point is that the I/O queue doesn't return any data itself, it just tells us that data is ready to be read.\nLastly it's only for our own bookkeeping we decrement the count of outstanding epoll_pending_events so we keep track of how many events we have pending.\n  Why even keep track of how many epoll_events are pending?  We don't use this value here, but I added it to make it easier to create some print statements showing the status of our runtime at different points. However, there are good reasons to keep track of these events even if we don't use them.\nOne area we're taking shortcuts on all the way here is security. If someone were to build a public facing server out of this, we need to account for slow networks and malicious users.\nSince we use IOCP, which is a completion based model, we allocate memory for a buffer for each Read or Write event. When we lend this memory to the OS, we're in a weird situation. We own it, but we can't touch it. There are several ways in which we could register interest in more events than occur, and thereby allocating memory that is just held in the buffers. Now if someone wanted to crash our server, they could cause this intentionally.\nA good practice is therefore to create a \"high watermark\" by keeping track of the number of pending events, and when we reach that watermark, we queue events instead of registering them with the OS.\nBy extension, this is also why you should  always  have a timeout on these events so that you at some point can reclaim the memory and return an timeout error if necessary.\n "

